From b872eda12d4ad146610e904e585b338fc9e8f944 Mon Sep 17 00:00:00 2001
From: Stella Laurenzo <stellaraccident@gmail.com>
Date: Fri, 21 Feb 2025 13:02:46 -0800
Subject: [PATCH 1/4] Use build RPATH correctly.

* Using BUILD_WITH_INSTALL_RPATH is *only* ever valid if the build and install trees share the same layout and both should have the same origin-relative setup. That is not the case here, and the proper thing to do is to rely on CMake to make sure that RPATHs for the build tree are absolute (which is its normal invariant) and to only use the install RPATH for install.
* Hard-coding of an LD_LIBRARY_PATH to a presumed installed ROCM is not allowed and is dangerous. It is also not necessary if the build RPATHs are managed by CMake vs being overriden.
* Even if the above were true, it is not portable and would need to be contained to just on Linux for a non empty ROCM_DIR.
---
 library/src/CMakeLists.txt | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/library/src/CMakeLists.txt b/library/src/CMakeLists.txt
index b625209d..587a04a0 100644
--- a/library/src/CMakeLists.txt
+++ b/library/src/CMakeLists.txt
@@ -75,7 +75,6 @@ target_include_directories( rocfft_rtc_helper
 set(APPEND_ROCMLIB_RPATH "\$ORIGIN/../../../lib")
 target_link_libraries( rocfft_rtc_helper PRIVATE rocfft-rtc-compile )
 set_target_properties( rocfft_rtc_helper PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON
-                                         BUILD_WITH_INSTALL_RPATH TRUE
                                          INSTALL_RPATH "${APPEND_ROCMLIB_RPATH}" )
 target_link_directories( rocfft_rtc_helper PRIVATE ${ROCFFT_HOST_LINK_DIRS} )
 
@@ -524,11 +523,9 @@ if ( ROCFFT_KERNEL_CACHE_ENABLE )
   list( REMOVE_ITEM GPU_TARGETS_AOT gfx1102 )
   list( REMOVE_ITEM GPU_TARGETS_AOT gfx1151 )
   list( REMOVE_ITEM GPU_TARGETS_AOT gfx1200 )
-  # The binary will be having relative RUNPATH with respect to install directory
-  # Set LD_LIBRARY_PATH for executing the binary from build directory.
   add_custom_command(
     OUTPUT rocfft_kernel_cache.db
-    COMMAND ${CMAKE_COMMAND} -E env "LD_LIBRARY_PATH=$ENV{LD_LIBRARY_PATH}:${ROCM_PATH}/${CMAKE_INSTALL_LIBDIR}" ./rocfft_aot_helper \"${ROCFFT_BUILD_KERNEL_CACHE_PATH}\" ${ROCFFT_KERNEL_CACHE_PATH} $<TARGET_FILE:rocfft_rtc_helper> ${GPU_TARGETS_AOT}
+    COMMAND "${CMAKE_CURRENT_BINARY_DIR}/rocfft_aot_helper" \"${ROCFFT_BUILD_KERNEL_CACHE_PATH}\" ${ROCFFT_KERNEL_CACHE_PATH} $<TARGET_FILE:rocfft_rtc_helper> ${GPU_TARGETS_AOT}
     DEPENDS rocfft_aot_helper rocfft_rtc_helper
     COMMENT "Compile kernels into shipped cache file"
   )
-- 
2.47.1.windows.2

